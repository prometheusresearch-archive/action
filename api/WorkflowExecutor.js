/**
 * @flow
 */

import * as React from 'react';
const mkDebug = require('debug');
const invariant = require('invariant');

const debug = mkDebug('RexAction:api:WorkflowExecutor');

import type {Action, Context, ContextTypeShape, View, Query, Data} from './Workflow.js';

export type Config = {
  initialAction: Action,
  initialContext?: Context,
  waitForUserInput: (
    Context,
    Data,
    (Context) => void,
    (Context, Data, (Context) => void) => React.Element<*>,
  ) => *,
  waitForData: (Query<*>) => Promise<{get(mixed): mixed}>,
  onState?: State => void,
  onTransition?: StateTransition => void,
};

export type State = {
  action: Action,
  context: Context,

  bail: 'allow' | 'ignore',
};

export type StateTransition = Next | Bail | Noop;

type Next = {
  type: 'Next',
  state: State,
};

type Bail = {
  type: 'Bail',
};

type Noop = {
  type: 'Noop',
};

function bail(state: State): StateTransition {
  switch (state.bail) {
    case 'allow':
      return {type: 'Bail'};
    case 'ignore':
      return {type: 'Noop'};
    default:
      invariant(false, 'Impossible');
  }
}

function next(state: State): StateTransition {
  return {type: 'Next', state};
}

function nextWithContext(state: State, context: Context): StateTransition {
  return {type: 'Next', state: {...state, context}};
}

export async function run({
  initialAction,
  initialContext,
  waitForUserInput,
  waitForData,
  onState,
  onTransition,
}: Config) {
  async function fetchData(state: State) {
    function findQuery(state: State) {
      const {action, context} = state;
      switch (action.type) {
        case 'View': {
          return action.query(context);
        }
        case 'Process': {
          return action.query(context);
        }
        case 'Guard': {
          return action.query(context);
        }
        case 'Sequence': {
          return null;
        }
        case 'Choice': {
          return null;
        }
        default:
          invariant(false, 'Unknown action: %s', action.type);
      }
    }

    const query = findQuery(state);
    debug('query', query);

    if (query != null) {
      const data = await waitForData(query);
      return data;
    } else {
      return {};
    }
  }

  async function step(state: State): Promise<StateTransition> {
    debug('state', state);
    if (onState != null) {
      onState(state);
    }
    const transition = await stepImpl(state);
    debug('transition', transition);
    if (onTransition != null) {
      onTransition(transition);
    }
    return transition;
  }

  async function stepImpl(state: State): Promise<StateTransition> {
    const {context, action} = state;

    if (!contextConformsTo(context, getContextRequirements(action))) {
      return bail(state);
    }

    const data = await fetchData(state);

    switch (action.type) {
      case 'View': {
        return new Promise(resolve => {
          const onContext = context => {
            state = {...state, bail: 'ignore'};
            resolve(nextWithContext(state, context));
          };
          waitForUserInput(state.context, data, onContext, action.render);
        });
      }

      case 'Process': {
        const nextContext = await action.execute(state.context, data);
        return nextWithContext(state, nextContext);
      }

      case 'Guard': {
        const isAllowed: boolean = Boolean(data);
        if (isAllowed) {
          return next(state);
        } else {
          return bail(state);
        }
      }

      case 'Sequence': {
        let nextState = {...state};
        for (const childAction of action.sequence) {
          nextState = {...nextState, action: childAction};
          const next: StateTransition = await step(nextState);
          if (next.type === 'Next') {
            nextState = next.state;
            continue;
          } else {
            return next;
          }
        }
        return next(nextState);
      }

      case 'Choice': {
        for (const childAction of action.choice) {
          const scopedState = {...state, bail: 'allow', action: childAction};
          const next: StateTransition = await step(scopedState);
          if (next.type === 'Bail') {
            continue;
          } else {
            return next;
          }
        }
        return next(state);
      }

      default:
        invariant(false, 'Unknown action: %s', action.type);
    }
  }

  const state: State = {
    prev: null,
    action: initialAction,
    context: initialContext || {},
    bail: 'ignore',
  };

  return await step(state);
}

function contextConformsTo(context: Context, shape: ContextTypeShape) {
  for (const key in shape) {
    const value = context[key];
    if (value == null) {
      return false;
    }
    const type = shape[key];
    switch (type.type) {
      case 'NumberType':
        if (typeof value !== 'number') {
          return false;
        }
        break;
      case 'StringType':
        if (typeof value !== 'string') {
          return false;
        }
        break;
      case 'EntityType':
        if (!(typeof value === 'object' && value != null && value.__type === type.name)) {
          return false;
        }
        break;
      default:
        // TODO:
        invariant(false, 'Need to handle type: %s', type.type);
    }
  }
  return true;
}

function getContextRequirements(action: Action) {
  switch (action.type) {
    case 'View':
    case 'Process':
    case 'Guard':
      return action.requires;
    case 'Sequence':
    case 'Choice':
      return {};
    default:
      invariant(false, 'Unknown action: %s', action.type);
  }
}
