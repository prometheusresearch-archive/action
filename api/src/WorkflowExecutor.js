/**
 * @flow
 */

const invariant = require('invariant');

import type {Action, Context, View} from './Workflow.js';

type State = {
  prev: ?State,
  action: Action,
  context: Context,
};

type WorkflowEvent = RenderView | EndOfWorkflow;

type RenderView = {
  type: 'RenderView',
  view: View<*>,
};

type EndOfWorkflow = {
  type: 'EndOfWorkflow',
};

export function run(
  initialAction: Action,
  initialContext: Context,
  handle: WorkflowEvent => void,
) {
  function aux(state) {
    const {action, context} = state;
    switch (action.type) {
      case 'View': {
      }
      case 'Process': {
      }
      case 'Guard': {
      }
      case 'Sequence': {
      }
      case 'Choice': {
      }
      default:
        invariant(false, 'Unknown action: %s', action.type);
    }
  }

  const state: State = {prev: null, action: initialAction, context: initialContext};
  aux(state);
}
