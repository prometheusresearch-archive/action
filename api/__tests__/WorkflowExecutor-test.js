/**
 * @flow
 */

import * as React from 'react';
import type {Action, Context} from '../Workflow.js';
import {type StateTransition, run} from '../WorkflowExecutor.js';

async function runAndCaptureTrace({
  initialAction,
  initialContext = {},
}: {
  initialAction: Action,
  initialContext?: Context,
}) {
  const trace = [];

  function onState(state) {
    trace.push({event: 'state', action: state.action.id, context: state.context});
  }

  function onTransition(transition: StateTransition) {
    trace.push({event: 'transition', type: transition.type});
  }

  async function waitForUserInput(state, data, onContext, render) {
    trace.push({event: 'render', context: state.context});
    render(state.context, data, onContext);
  }

  function waitForData(query) {
    trace.push({event: 'fetch', query});
    return Promise.resolve(new Map());
  }

  const next: ?StateTransition = await run({
    initialAction,
    initialContext,
    waitForUserInput,
    waitForData,
    onState,
    onTransition,
  });

  return {trace};
}

function view(params): Action {
  return {
    type: 'View',
    id: params.id,
    requires: params.requires || {},
    provides: {},
    query(context) {
      return '';
    },
    render(context, data, onContext) {
      if (params.onContext != null) {
        onContext(params.onContext(context));
      } else {
        onContext(context);
      }
      return <div>Hello</div>;
    },
  };
}

function set(key, value) {
  return context => ({...context, [key]: value});
}

test('one action workflow', async function() {
  const initialAction = view({id: 'hello', onContext: set('a', 43)});

  const {trace} = await runAndCaptureTrace({initialAction});
  expect(trace).toMatchSnapshot();
});

test('one action workflow bail', async function() {
  const initialAction = view({
    id: 'hello',
    requires: {
      a: {type: 'StringType'},
    },
    onContext: set('a', 43),
  });

  {
    const {trace} = await runAndCaptureTrace({initialAction});
    expect(trace).toMatchSnapshot();
  }

  {
    const {trace} = await runAndCaptureTrace({initialAction, initialContext: {a: 'ok'}});
    expect(trace).toMatchSnapshot();
  }

  {
    const {trace} = await runAndCaptureTrace({initialAction, initialContext: {a: 21}});
    expect(trace).toMatchSnapshot();
  }
});

test('two action sequence workflow', async function() {
  const initialAction = {
    type: 'Sequence',
    id: 'sequence',
    sequence: [
      view({id: 'hello', onContext: set('a', 43)}),
      view({id: 'world', onContext: set('b', 44)}),
    ],
  };

  const {trace} = await runAndCaptureTrace({initialAction});
  expect(trace).toMatchSnapshot();
});

test('two action sequence workflow - bail on second', async function() {
  const initialAction = {
    type: 'Sequence',
    id: 'sequence',
    sequence: [
      view({
        id: 'hello',
      }),
      view({
        id: 'world',
        requires: {
          req: {type: 'StringType'},
        },
      }),
    ],
  };

  const {trace} = await runAndCaptureTrace({initialAction});
  expect(trace).toMatchSnapshot();
});

test('two action choice workflow', async function() {
  const initialAction: Action = {
    type: 'Choice',
    id: 'choice',
    choice: [
      view({id: 'hello', onContext: set('a', 43)}),
      view({id: 'world', onContext: set('b', 44)}),
    ],
  };

  const {trace} = await runAndCaptureTrace({initialAction});
  expect(trace).toMatchSnapshot();
});
