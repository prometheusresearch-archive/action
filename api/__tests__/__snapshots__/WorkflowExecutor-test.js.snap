// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one action workflow 1`] = `
Array [
  Object {
    "action": "hello",
    "context": Object {},
    "event": "state",
  },
  Object {
    "event": "fetch",
    "query": Object {},
  },
  Object {
    "context": Object {},
    "event": "render",
  },
  Object {
    "event": "transition",
    "type": "Next",
  },
]
`;

exports[`one action workflow bail 1`] = `
Array [
  Object {
    "action": "hello",
    "context": Object {},
    "event": "state",
  },
  Object {
    "event": "transition",
    "type": "Noop",
  },
]
`;

exports[`one action workflow bail 2`] = `
Array [
  Object {
    "action": "hello",
    "context": Object {
      "a": "ok",
    },
    "event": "state",
  },
  Object {
    "event": "fetch",
    "query": Object {},
  },
  Object {
    "context": Object {
      "a": "ok",
    },
    "event": "render",
  },
  Object {
    "event": "transition",
    "type": "Next",
  },
]
`;

exports[`one action workflow bail 3`] = `
Array [
  Object {
    "action": "hello",
    "context": Object {
      "a": 21,
    },
    "event": "state",
  },
  Object {
    "event": "transition",
    "type": "Noop",
  },
]
`;

exports[`two action choice workflow 1`] = `
Array [
  Object {
    "action": "choice",
    "context": Object {},
    "event": "state",
  },
  Object {
    "action": "hello",
    "context": Object {},
    "event": "state",
  },
  Object {
    "event": "fetch",
    "query": Object {},
  },
  Object {
    "context": Object {},
    "event": "render",
  },
  Object {
    "event": "transition",
    "type": "Next",
  },
  Object {
    "event": "transition",
    "type": "Next",
  },
]
`;

exports[`two action sequence workflow - bail on second 1`] = `
Array [
  Object {
    "action": "sequence",
    "context": Object {},
    "event": "state",
  },
  Object {
    "action": "hello",
    "context": Object {},
    "event": "state",
  },
  Object {
    "event": "fetch",
    "query": Object {},
  },
  Object {
    "context": Object {},
    "event": "render",
  },
  Object {
    "event": "transition",
    "type": "Next",
  },
  Object {
    "action": "world",
    "context": Object {},
    "event": "state",
  },
  Object {
    "event": "transition",
    "type": "Noop",
  },
  Object {
    "event": "transition",
    "type": "Noop",
  },
]
`;

exports[`two action sequence workflow 1`] = `
Array [
  Object {
    "action": "sequence",
    "context": Object {},
    "event": "state",
  },
  Object {
    "action": "hello",
    "context": Object {},
    "event": "state",
  },
  Object {
    "event": "fetch",
    "query": Object {},
  },
  Object {
    "context": Object {},
    "event": "render",
  },
  Object {
    "event": "transition",
    "type": "Next",
  },
  Object {
    "action": "world",
    "context": Object {
      "a": 43,
    },
    "event": "state",
  },
  Object {
    "event": "fetch",
    "query": Object {},
  },
  Object {
    "context": Object {
      "a": 43,
    },
    "event": "render",
  },
  Object {
    "event": "transition",
    "type": "Next",
  },
  Object {
    "event": "transition",
    "type": "Next",
  },
]
`;
